File: Dogs-C/Dogs-C/View Controllers/JLBreedListTableViewController.m
//
//  JLBreedListTableViewController.m
//  Dogs-C
//
//  Created by Jordan Lamb on 10/9/19.
//  Copyright © 2019 Squanto Inc. All rights reserved.
//

#import "JLBreedListTableViewController.h"
#import "JLBreedController.h"
#import "JLSubBreedsTableViewController.h"
#import "JLImageCollectionViewController.h"

@interface JLBreedListTableViewController ()

@property (nonatomic, copy) NSArray *breeds;

@end

@implementation JLBreedListTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [[JLBreedController sharedInstance] fetchBreeds:^(NSArray *breeds) {
        self.breeds = breeds;
        dispatch_async(dispatch_get_main_queue(), ^{
            [[self tableView] reloadData];
        });
    }];
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.breeds.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"breedCell" forIndexPath:indexPath];
    cell.textLabel.text = [[self.breeds[indexPath.row] name] capitalizedString];;
    return cell;
}


#pragma mark - Navigation

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    JLBreed *breed = self.breeds[[[[self tableView] indexPathForSelectedRow] row]];
    if( [breed.subBreeds count] > 0)
    {
        [self performSegueWithIdentifier:@"toSubBreedVC" sender:self];
    } else {
        [self performSegueWithIdentifier:@"toImagesVC" sender:self];
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    JLBreed * breed = self.breeds[[[[self tableView] indexPathForSelectedRow] row]];
    if([segue.identifier  isEqualToString: @"toSubBreedVC"])
    {
        //If it's to the SubBreedTableViewController. Send the breed to that landing pad.
        JLSubBreedsTableViewController *destinationVC = segue.destinationViewController;
        destinationVC.breed = breed;
        
    }else if([segue.identifier  isEqualToString: @"toImagesVC"]){
        JLImageCollectionViewController *destinationVC = segue.destinationViewController;
        destinationVC.breed = breed;
    }
}

@end

File: Dogs-C/Dogs-C/View Controllers/JLBreedListTableViewController.swift
//
//  JLBreedListTableViewController.swift
//  Dogs-C
//
//  Created by Jordan Lamb on 10/9/19.
//  Copyright © 2019 Squanto Inc. All rights reserved.
//

import UIKit

class JLBreedListTableViewController: UITableViewController {
    
    var fetchedBreeds: [JLBreed] = []

    override func viewDidLoad() {
        super.viewDidLoad()
        JLBreedController.sharedInstance().fetchBreeds { (breed) in
            self.fetchedBreeds = breed
            DispatchQueue.main.async {
                self.tableView.reloadData()
            }
        }
    }

    // MARK: - Table view data source
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return fetchedBreeds.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "breedCell", for: indexPath)
        let breed = fetchedBreeds[indexPath.row]
        cell.textLabel?.text = breed.name
        return cell
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let breed = self.fetchedBreeds[indexPath.row]
        if breed.subBreeds.count > 0 {
            performSegue(withIdentifier: "toSubBreedVC", sender: self)
        } else {
            performSegue(withIdentifier: "toImagesVC", sender: self)
        }
    }
    // MARK: - Navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "toSubBreedVC" {
            guard let index = tableView.indexPathForSelectedRow else { return }
            guard let destination = segue.destination as? JLSubBreedTableViewController else { return }
            let objectToSend = fetchedBreeds[index.row]
            destination.breeds = objectToSend
        } else if segue.identifier == "toImagesVC" {
            guard let index = tableView.indexPathForSelectedRow else { return }
            guard let destination = segue.destination as? JLImageCollectionViewController else { return }
            let objectToSend = fetchedBreeds[index.row]
            destination.breeds = objectToSend
        }
    }

}
